transformer_engine/pytorch/float8_tensor.py

```diff
@@ -0,0 +1,654 @@
+                )
+            return super().transpose(dim0, dim1)
+
+        if not cache:
+            # Use optimized kernel for basic 2D transpose
+            # TODO Support differentiation # pylint: disable=fixme
+            return Float8Tensor.make_like(
+                self,
+                data=tex.fp8_transpose(
+                    self._data.contiguous().detach(),
+                    self._fp8_dtype,
+                ),
+            )
+
+        # Handle caching
+        if self._transpose is None:
+            self._transpose = self.transpose()
+        return self._transpose
+
+    @torch.no_grad()
+    def reset_fp8_meta_scale_inv(self) -> None:
+        """Replace FP8 meta tensor scale-inverse with cached value
+
+        The FP8 meta tensor scale_inv entry corresponding to this
+        tensor is replaced with the scale_inv value used to construct
+        the tensor.
+
+        """
+        if self._fp8_meta is None:
+            return
+        fp8_meta_key = FP8GlobalStateManager.get_meta_tensor_key(
+            forward=self._fp8_meta_forward,
+        )
+        scale_inv = self._fp8_meta[fp8_meta_key].scale_inv[self._fp8_meta_index]
+        scale_inv.copy_(self._scale_inv)
+
+    def to_dtype(self, dtype: torch.dtype) -> Float8Tensor:
+        """Create `Float8Tensor` with given nominal dtype
+
+        The new tensor has the same underlying FP8 data.
+
+        """
+        return Float8Tensor.make_like(self, data=self._data, dtype=dtype)
+
+    def _reset_caches(self) -> None:
+        """Reset cached values
+
+        Should be called after any in-place operation.
+
+        """
+        self._transpose = None
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
+
+        # In-place copy op
+        if func == aten.copy_.default:
+
+            # Check tensors
+            dst = args[0]
+            src = args[1]
+            if not isinstance(dst, Float8Tensor):
+                raise RuntimeError("Expected to copy into Float8Tensor")
+            if not isinstance(src, torch.Tensor):
+                raise RuntimeError("Expected to copy from tensor")
+            if not dst._data.is_contiguous():
+                raise RuntimeError("Transformer Engine cast kernels require contiguous data")
+
+            # Make sure input is in expected format
+            if isinstance(src, Float8Tensor):
+                src = src.from_float8()
+            src = src.expand(dst.size())
+            src = src.to(
+                device=dst.device,
+                memory_format=torch.contiguous_format,
+            )
+
+            # Cast to FP8
+            tex.cast_to_fp8_noalloc(
+                src.view(1,-1),
+                dst._scale_inv.reciprocal(),
+                dst._data.view(1,-1),
+                torch.empty_like(dst._scale_inv),  # amax
+                dst._scale_inv,
+                dst._fp8_dtype,
+            )
```

Comment:
I've noticed we never update `scale_inv`, even if fp8_meta changes. I think updating within `copy_` should handle this:
```suggestion
            # Update scaling factor if FP8 meta tensors are available
            if dst._fp8_meta is None:
                scale = dst._scale_inv.reciprocal()
            else:
                fp8_meta_key = FP8GlobalStateManager.get_meta_tensor_key(
                    forward=dst._fp8_meta_forward,
                )
                scale = dst._fp8_meta[fp8_meta_key].scale[dst._fp8_meta_index]
                dst._scale_inv = scale.detach().view(1).reciprocal()

            # Cast to FP8
            tex.cast_to_fp8_noalloc(
                src.view(1,-1),
                scale,
                dst._data.view(1,-1),
                torch.empty_like(dst._scale_inv),  # amax
                dst._scale_inv,
                dst._fp8_dtype,
            )
```

Reply:
Ah right, there used to be a swapping logic for `scale_inv` when we first discussed this. 
Having it in `copy_` seems fine but I guess we might need it in other in-place operations like `add_`? So, we could have updating `scale_inv` as part of some "inplace_op_hook". Wdyt?

Reply:
The other in-place ops call `copy_` when casting the result back to FP8, so shouldn't `copy_` be sufficient?

Reply:
Yes `copy_` should be sufficient since it's the end point of inplace ops. Trying the approach of storing `_scale_inv` in a mutable data structure as opposed to a direct attribute in `Float8Tensor`.

Reply:
Okay `copy_` seems to be the only realized in-place op in Float8Tensor class, so this sounds good then!